{% import 'macros.html' as macros %}
{% extends "layout.html" %}

{% set active_page = "view_variants" %}
{% block header%}
<link href="/static/dashboard.css" rel="stylesheet" media="screen">
<link href="/static/custom.css" rel="stylesheet" media="screen">
<script src="/static/d3.v3/d3.v3.min.js"></script>
<script src="/static/dashboard.js"></script>
<script src="/static/gene_model.js"></script>
<link href="/static/gene_model.css" rel="stylesheet" media="screen">

<style>
.annotation-tabs li {font-size: 16px}
.annotation-tabs li.active {font-weight: bold}
</style>
{% endblock %}
{% block title %}
Genotype Details: chr{{ data["chrom"] }}: {{data["start"]}}{{ '-%d' % data.sto if data.stop }}
{% endblock %}

{% block content %}


<h3>Basic Information</h3>
<dl class="dl-horizontal">
  <dt>Position</dt>
    <dd>chr{{ data["chrom"] }}: {{data["start"]}}{{ '-%d' % data.sto if data.stop }} (<a href="{{ ucsc_linkout(data["chrom"], data["start"], data["end"]) }}" target="_blank">UCSC</a>, <a href="{{ esp_linkout(data["chrom"], data["start"], data["end"]) }}" target="_blank">ESP</a>)</dd>
  <dt>Sample ID</dt>
    <dd><a href='/samples/id:{{ data["sample_id"] }}'>{{ data["sample_name"] }}</a></dd>
  <dt>Alt / Ref</dt>
    <dd>{{ data["alt"]}}  / {{ data["ref"] }}</dd>
  {% if "filter" in data %}
  <dt>Filters/Sets</dt>
    <dd>{{ data["filter"] | join(', ')}}</dd>
  {% endif %}

  {% if "id" in data %}
    <dt>dbSNP</dt>
      <dd><a href="http://www.ncbi.nlm.nih.gov/projects/SNP/snp_ref.cgi?rs={{ data["id"]|replace('rs','')}}">{{ data["id"]|default("")}}</a></dd> 
  {% endif %}
</dl>


<h3>Annotations and Functional predictions</h3>
<div class="row-fluid">
  <div class="span8">
<table class="table">
  <thead>
    <th>Type</th><th>Gene</th><th>AA Change</th><th>Codon Change</th><th>Transcript</th><th>Exon #</th><th></th>
  </thead>
  <tbody>
    {% for row in data["annotations"]["EFF"] %}
      <tr>
        <td><span class="label impact-tag {{ row['effect_code'] }}">{{row["e"]}}</span></td>
        <td><a href="/variants/{{row["g"]}}">{{row["g"]}}</a></td>
        <td>{{row["aa"]}}</td>
        <td>{{row["cc"]}}</td>
        <td>{{row["tx"]}}</td>
        <td>{{row["r"]}}</td>
        <td><a href="#" onClick="show_geneplot('{{row["tx"]}}','{{row["aa"]}}');">View GenePlot</a></td>
      </tr>
    {% endfor %}
    </tbody>
  </table>
</div></div>


<ul class="nav nav-tabs annotation-tabs" id="myTab">
  <li class="active"><a href="#summary" data-toggle="tab">Dashboard</a></li>
  <li><a href="#gene_model" data-toggle="tab">GenePlot</a></li>
  <li><a href="#dbnsfp" data-toggle="tab">dbNSFP Annotations</a></li>
  <li><a href="#gatk" data-toggle="tab">GATK Annotations</a></li>
  
  <li><a href="#raw" data-toggle="tab">Raw Record</a></li>
</ul>
 
<div class="tab-content">
  <div class="tab-pane active" id="summary">
<!--<div style="padding-left: 10px"><img src="/static/mockup/dashboard_mockup.png" width="700px"/></div>-->
  <div id="div-dashboard"></div>
   
   {% if "dbNSFP" in data["annotations"]  %}    
    <script>
      score_data = {{ data["dashboard"]["scores"]|tojson|safe }}
      freq_data = {{ data["dashboard"]["freq"]|tojson|safe }}
      gatk_data = {{ data["dashboard"]["gatk"]|tojson|safe }}

      make_dashboard("#div-dashboard", score_data, freq_data, gatk_data);
    </script>
  {% else %}
     <em>No dbNSF annotations for this variant!</em>
  {% endif %}
  

  </div>
  <div class="tab-pane" id="dbnsfp">
    
      {% if "dbNSFP" in data["annotations"]  %}
        <dl class="dl-horizontal">
        {% for row in data["annotations"]["dbNSFP"] %}
          <dt>{{ row }}</dt><dd>{{- data["annotations"]["dbNSFP"][row] -}}</dd>
        {% endfor %}
        </dl>
      {% else %}
        <em>No dbNSF annotations for this variant!</em>
      {% endif %}
    
  </div>
  <div class="tab-pane" id="gatk">
    <dl class="dl-horizontal">
      {% for row in data["annotations"] %}
        {% if row not in ["EFF","dbNSFP"] %}
        <dt>{{ row }}</dt>
          <dd>{{ data["annotations"][row]}}</dd>
        {% endif %}
      {% endfor %}
    </dl>
  </div>
  <div class="tab-pane" id="gene_model">
    <div id="gene_model_div"></div>
    <script>
    var isoform = "{{data["annotations"]["EFF"][0]["tx"] }}";
    var gene_model = d3.select("#gene_model_div").
                append("svg:svg").
                attr("id","gene_model_svg").
                attr("width", 900).
                attr("height", 300);
    function show_geneplot(isoform, position){
        variants_positions = position.match(numberPattern);
        $.getJSON("/uniprot/"+isoform+".json", function(d) {
                   data = d;
                   domains = []
                   features = d["result"][0]["features"];
                   for (var i = 0; i < features.length; i++) {
                        if (PLOT_FEATURES.indexOf(features[i][0]) >= 0){
                            domains.push({"start": features[i][1], 
                                          "stop": features[i][2],
                                          "type": features[i][0],
                                          "name": features[i][3]
                                        });
                        }
                   }
                   aa_length = d["result"][0]["sequence_length"]
                   console.log(aa_length)
                   draw_model(gene_model, domains, aa_length);
                   draw_variants(variants_positions)
                });
        $('#myTab a[href="#gene_model"]').tab('show')
    };  

    var data;
    var domains;
    var eff_data = {{ data["annotations"]["EFF"]|tojson|safe }}
    var numberPattern = /\d+/g;
    // variants_positions = []
    // for (var i=0; i < eff_data.length; i++){
    //     variants_positions = variants_positions.concat(eff_data[i]["aa"].match(numberPattern))
    // }
    //show_geneplot(isoform, variants_positions)
    
    </script>
  </div>
  <div class="tab-pane" id="raw">
    <h3>Record</h3>
    {{ data | pprint(verbose=True)}}
    {% endblock %}
  </div>
</div>








{% macro esp_linkout(chrom, start, stop) -%}
    {% if stop == None %}
      http://evs.gs.washington.edu/EVS/PopStatsServlet?searchBy=chromosome&chromosome={{ chrom }}&chromoStart={{ start }}&chromoEnd={{ start + 1 }}
    {% else %}
    http://evs.gs.washington.edu/EVS/PopStatsServlet?searchBy=chromosome&chromosome={{ chrom }}&chromoStart={{ start }}&chromoEnd={{ stop }}
    {% endif %}
    
{%- endmacro %}

{% macro ucsc_linkout(chrom, start, stop) -%}
    {% if stop == None %}
      http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&org=human&position=chr{{ chrom }}%3A{{ start }}-{{ start + 1}}
    {% else %}
      http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&org=human&position=chr{{ chrom }}%3A{{ start }}-{{ stop }}
    {% endif %}
    
{%- endmacro %}